binfetcher: generate a bash script to generate URL

tbg opened the pull request:
This isn't immediately 100% useful as it still requires handling the
decompression manually. However, I wanted to try out the approach; it's not too
hard to add decompression by matching on the $suffix variable.

There's something off-putting about generating bash from a Go script, but this
seems better than having two distinct versions due to the coupling between them
in both code and tests. If we decide that this is a bad idea at some point, it's
easy to shell out the script into a standalone entity.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/21054)
<!-- Reviewable:end -->

tbg committed:
binfetcher: generate a bash script to generate URL

This isn't immediately 100% useful as it still requires handling the
decompression manually. However, I wanted to try out the approach; it's not too
hard to add decompression by matching on the $suffix variable.

There's something off-putting about generating bash from a Go script, but this
seems better than having two distinct versions due to the coupling between them
in both code and tests. If we decide that this is a bad idea at some point, it's
easy to shell out the script into a standalone entity.

Release note: None

knz commented on pkg/util/binfetcher/binfetcher.go:
```diff
@@ -44,20 +54,154 @@ type Options struct {
+binary="%s"
+version="%s"
+os=""
+arch="amd64" # TODO(tschottdorf): how to auto-detect?
```
`$(uname -m)`

knz replied:
However since these scripts are invoked from Go code I highly recommend you to pass the arch as argument from Go (you probably want the same arch as the cockroachdb binary, i.e. the string from the `build` package.

tschottdorf replied:
The script shouldn't be tied to the arch on which it was generated. The idea is that this can be used in CI for various auxiliary components that need to run a cluster of a certain version (if I understood @benesch correctly). So it wants to just "get the right one for the system on which the script is called".

Also, on my MacBook `uname -m` results in `x86_64`.

knz replied:
oh that's unfortunate.

My personal preference would be to name the binaries after the output of `uname -m` on each platform (because that's stable and reliable!) not what Go thinks it should be (that's usually... ad-hoc, if not to say bogus).

Oh well.

However my other proposal remains -- compute the string in Go and bake it in the script by string substitution?

benesch replied:
What about `go env GOARCH`?

benesch replied:
mmm I guess nothing else in this script depends on Go being available. Never mind.
